pkg/services/hcloud/server/server.go

```diff
@@ -925,3 +1432,67 @@ func updateHCloudMachineStatusFromServer(hm *infrav1.HCloudMachine, server *hclo
 	hm.Status.Addresses = statusAddresses(server)
 	hm.Status.InstanceState = ptr.To(server.Status)
 }
+
+func (s *Service) getSSHClient(ctx context.Context) (sshclient.Client, error) {
+	hm := s.scope.HCloudMachine
+
+	robotSecretName := s.scope.HetznerCluster.Spec.SSHKeys.RobotRescueSecretRef.Name
+	if robotSecretName == "" {
+		msg := "HetznerCluster.Spec.SSHKeys.RobotRescueSecretRef.Name is empty. Can not set hcloud machine into rescue-mode without ssh private key"
+		err := errors.New(msg)
+		conditions.MarkFalse(hm, infrav1.ServerAvailableCondition,
+			string(hm.Status.BootState), clusterv1.ConditionSeverityWarning,
+			"%s", msg)
+		return nil, err
+	}
+
+	robotSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      robotSecretName,
+			Namespace: s.scope.Namespace(),
+		},
+	}
+
+	err := s.scope.Client.Get(ctx, client.ObjectKeyFromObject(robotSecret), robotSecret)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			errReader := s.scope.APIReader.Get(ctx, client.ObjectKeyFromObject(robotSecret), robotSecret)
+			if errReader == nil {
+				msg := fmt.Sprintf("secret %q is missing label %s=%s", robotSecretName,
+					secretutil.LabelEnvironmentName, secretutil.LabelEnvironmentValue)
+				s.scope.Logger.Error(nil, msg)
+				conditions.MarkFalse(hm, infrav1.ServerAvailableCondition,
+					string(hm.Status.BootState), clusterv1.ConditionSeverityWarning,
+					"%s", msg)
+				return nil, errors.New(msg)
+			}
+		}
+		return nil, fmt.Errorf("failed to get secret %q: %w", robotSecretName, err)
+	}
+
+	if len(hm.Status.Addresses) == 0 {
+		msg := "HCloudMachine.Status.Addresses empty. Can not connect via ssh"
```

Comment:
sounds to me permanent.. If so, we shouldn't return the error

Reply:
This code is inside getSSHClient() which has this signature:

```go 
func (s *Service) getSSHClient(ctx context.Context) (sshclient.Client, error) 
```

Currently, there is no differentiation between a temporary and a permanent error.

This method gets called at two places.

I see these options:

- leave it like it is
- add a reconcile.Result to the return values
- create a ProvisioningFailedError and return that. In defer() we catch that error and then do SetError. (one place to handle it)
- create a sentinel error for this case, and let the caller handle that (two places to handle it).


@janiskemper what do you prefer?

Reply:
Are you sure? This has never happened up to now. So would just leave the code like it is. Handling something in the main logic which never happens.... Is that your intention?